lib/scripting/grammar/grammar.ts

```diff
@@ -0,0 +1,268 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import { Grammars, IToken, Parser } from 'ebnf';
+import { generate } from 'escodegen';
+import { Program, Statement } from 'estree';
+import { getTextFile } from '../../refs.node';
+import { program } from '../estree';
+import { ppArray } from '../post-process/array';
+import { ppAssign } from '../post-process/assign';
+import { ppBranch } from '../post-process/branch';
+import { ppCall } from '../post-process/call';
+import { ppClass } from '../post-process/class';
+import { ppConditional } from '../post-process/conditional';
+import { ppConst } from '../post-process/const';
+import { ppError } from '../post-process/error';
+import { ppExpr } from '../post-process/expr';
+import { ppHelpers } from '../post-process/helpers';
+import { ppLiteral } from '../post-process/literal';
+import { ppLoop } from '../post-process/loop';
+import { ppMethod } from '../post-process/method';
+import { ppVarDecl } from '../post-process/vardecl';
+import { ppWith } from '../post-process/with';
+
+const dashAst = require('dash-ast');
+
+export interface ESIToken extends IToken {
+	estree: any;
+	parent: ESIToken;
+	children: ESIToken[];
+}
+
+export class Grammar {
+	private TOKEN_TERMINAL_KEYWORDS = 'Keywords ::= ';
+
+	private GRAMMAR_TARGET_FORMAT = 'Format';
+	private GRAMMAR_TARGET_PROGRAM = 'Program';
+
+	private parser: Parser;
+	private keywords: { [index: string]: string } = {};
+
+	private postProcessors = [
+		ppHelpers,
+		ppLiteral,
+		ppExpr,
+		ppVarDecl,
+		ppConst,
+		ppArray,
+		ppError,
+		ppAssign,
+		ppConditional,
+		ppLoop,
+		ppBranch,
+		ppWith,
+		ppMethod,
+		ppCall,
+		ppClass,
+	];
+
+	constructor() {
+		let grammar = getTextFile('grammar.bnf');
+
+		this.setKeywords(grammar);
+
+		grammar = this.addCaseInsensitiveKeywords(grammar);
+
+		this.parser = new Parser(Grammars.Custom.getRules(grammar), {});
+	}
+
+	public transpile(script: string): Program {
+		const stmts: Statement[] = [];
+
+		const formattedScript = this.format(script);
+
+		const vbsAst = this.parser.getAST(formattedScript, this.GRAMMAR_TARGET_PROGRAM);
+
+		if (vbsAst === null) {
+			throw new Error('Unable to transpile script:\n\n' + formattedScript);
+		}
+
+		const postProcessors = this.postProcessors;
+
+		dashAst(vbsAst, {
+			leave(node: ESIToken, parent: ESIToken) {
+				if (node.type === 'Program') {
+					for (const child of node.children) {
+						if (child.estree) {
+							if (!Array.isArray(child.estree)) {
+								stmts.push(child.estree);
+							} else {
+								for (const statement of child.estree as Statement[]) {
+									stmts.push(statement);
+								}
+							}
+						}
+					}
+				} else {
+					let estree: any = null;
+					for (const postProcessor of postProcessors) {
+						estree = postProcessor(node);
+						if (estree) {
+							break;
+						}
```

Comment:
Correct.  I added a comment to explain this:

```
/**
 * Loop through all registered post processors until an estree
 * is returned. If no post processors can handle the node, and
 * the node has a child, copy it's estree. This will allow estrees
 * to get propagated to parents, and we wont need to have
  * post-processors for every single rule in the grammar.
 */
```

The big thing is the last point: `we wont need to have post-processors for every single rule in the grammar`

Reply:
Thanks 👍